pkg/build/gobuild.go

```diff
@@ -342,6 +338,26 @@ func (g *gobuild) tarKoData(importpath string) (*bytes.Buffer, error) {
 		_, err = io.Copy(tw, file)
 		return err
 	})
+}
+
+func (g *gobuild) tarKoData(importpath string) (*bytes.Buffer, error) {
+	buf := bytes.NewBuffer(nil)
+	// Compress this before calling tarball.LayerFromOpener, since it eagerly
+	// calculates digests and diffids. This prevents us from double compressing
+	// the layer when we have to actually upload the blob.
+	//
+	// https://github.com/google/go-containerregistry/issues/413
+	gw, _ := gzip.NewWriterLevel(buf, gzip.BestSpeed)
+	defer gw.Close()
+	tw := tar.NewWriter(gw)
+	defer tw.Close()
+
+	root, err := g.kodataPath(importpath)
+	if err != nil {
+		return nil, err
+	}
+
+	err = walkRecursive(tw, root, kodataRoot)
```

Comment:
why not just `return buf, err`?

Reply:
Doing that without copies sounds like a PITA with recursion.  This way the N calls can write to the same `buf` directly.

Reply:
I think he means you have redundant error checking here when you could just do:
```go
return buf, walkRecursive(tw, root, kodataRoot)
```

Reply:
done